filters/SkewnessBalancingFilter.cpp

```diff
@@ -0,0 +1,137 @@
+/******************************************************************************
+ * Copyright (c) 2019, Bradley J Chambers (brad.chambers@gmail.com)
+ *
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following
+ * conditions are met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in
+ *       the documentation and/or other materials provided
+ *       with the distribution.
+ *     * Neither the name of Hobu, Inc. or Flaxen Geo Consulting nor the
+ *       names of its contributors may be used to endorse or promote
+ *       products derived from this software without specific prior
+ *       written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
+ * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
+ * OF SUCH DAMAGE.
+ ****************************************************************************/
+
+#include "SkewnessBalancingFilter.hpp"
+
+#include <pdal/PointViewIter.hpp>
+#include <pdal/util/ProgramArgs.hpp>
+
+namespace pdal
+{
+
+static PluginInfo const s_info{
+    "filters.skewnessbalancing", "Bartels & Wei Skewness Balancing",
+    "http://pdal.io/stages/filters.skewnessbalancing.html"};
+
+CREATE_STATIC_STAGE(SkewnessBalancingFilter, s_info)
+
+std::string SkewnessBalancingFilter::getName() const
+{
+    return s_info.name;
+}
+
+void SkewnessBalancingFilter::addDimensions(PointLayoutPtr layout)
+{
+    layout->registerDim(Dimension::Id::Classification);
+}
+
+std::set<PointId> SkewnessBalancingFilter::processGround(PointViewPtr view)
+{
+    auto cmp = [](const PointIdxRef& p1, const PointIdxRef& p2) {
+        return p1.compare(Dimension::Id::Z, p2);
+    };
+    std::sort(view->begin(), view->end(), cmp);
+
+    point_count_t n(0);
+    point_count_t n1(0);
+    double delta, delta_n, term1, M1, M2, M3;
+    M1 = M2 = M3 = 0.0;
+    std::map<PointId, double> skewness;
```

Comment:
I don't understand why this is a map.  Unless I'm missing something, skewness is always accessed by pointId, so PointIdList would work.

Reply:
I probably changed course along the way. That said, it's not a PointIdList, as we are actually interested in the skewness, which is a double. I think vector<double> will do the trick.